app/models/concerns/smart_proxy_host_extensions.rb

```diff
@@ -0,0 +1,52 @@
+module SmartProxyHostExtensions
+  extend ActiveSupport::Concern
+
+  module ClassMethods
+    def smart_proxy_reference(hash)
+      SmartProxyReference.add_smart_proxy_reference hash
+    end
+
+    def smart_proxy_ids(hosts_scope = Host::Managed.where(nil))
+      hosts_scope.eager_load(proxy_join_tables).pluck(proxy_column_list).flatten.uniq.compact
+    end
+
+    def proxy_self_columns
+      proxy_join_references(true).values.first.map(&:to_s)
+    end
+
+    def proxy_join_tables
+      proxy_join_references.keys
+    end
+
+    def create_groups
+      refs = proxy_join_references.inject([]) { |memo, (k, v)| memo << { k => v } }
+      refs.select { |ref| host_reflection(ref.keys.first) }
+        .group_by { |h| host_reflection(h.keys.first).table_name }.with_indifferent_access
```

Comment:
The `ref` argument in the line 23 and `h` in 24 represent the same type of items, right? Using different names makes it harder to read the code